masschallenge/impact-api
# AC-4501: Add v1 API to support RESTful user calls

To test:

1) Get the branch.

2) make dev

3) `make dbload GZ_FILE=../accelerate/test_data/initial_schema.sql.gz`

4) Grant v1_clients permission to a user without any existing permissions and answer 'y' when/if it asks.  E.g.,
```
$ make grant-permissions PERMISSION_USER=37891-user@example.com PERMISSION_CLASSES=v1_clients
Create group name v1_clients? y
granting 37891-user@example.com add to group v1_clients
```

5) Login to the local impact-api server as the given user.

6) Create an application for the user.  Provide name, select Client type => Public, Authorization grant type => Resource owner password-based, click 'Save'.

7) In Postman load the [collection](https://www.getpostman.com/collections/4b28d750ca0b7d27dc77):  File => Import => Import from Link

8) Set the Environment variables for the user and the client id (<eye> => Edit)

9) Send 'oauth token' API from Postman

10)   Send 'v1 user list get' (should see a list of user info in the results).

11) You can test the next and previous links by pasting the URL into the url box next to 'GET V'.  This should behave the same way that they do for more generic DRF API calls like 'startup list'.  Running 'v1 user list get' should set the Postman `variable random_user_id` which is used by the other 'v1 user get' and 'v1 user patch'.

**Yotam: As far as I can see the script that sets `random_user_id` is at "simple user list". I had to add it manually to "v1 user list get", and then it worked as described.**
**Nathan: That was true for an earlier version of the collection, but it should be updated now.  You may need to reload the collection.  Let me know if that doesn't work since that would mean there's something I don't understand about Postman.**

12) You should now be able to test 'v1 user get' and 'v1 user patch'.  Note if you check off items in 'v1 user patch' it should succeed and only update the appropriate fields when you then run 'v1 user get'.

13) POST should work as well and again will set the `random_user_id` appropriately.  If you send the same args more than once you should get an error since the email is already in use.  If you are missing requirement arguments it should complain.  `is_active` is the only optional argument.

To locally test that the users created by the API work in the accelerate platform you need to first connect your local impact-api server to the MySQL instance running on your vagrant machine. Unfortunately this is pretty tricky to do at the moment, but the follow instruction should get you there:

1) Switch your local accelerate repo to the 'development' branch and run `resetdb`.

2) Edit your Vagrantfile and add the following line:
```
      mc_18.vm.network :forwarded_port, guest: 3306, host: 8306, auto_correct: true
```

after the line:
```
       mc_18.vm.network :forwarded_port, guest: 8111, host: 8181, auto_correct: true
```

Note that this is the line that starts with `mc_18.` not `mc_18_safe.`.  More precisely:
```
200a201
>       mc_18.vm.network :forwarded_port, guest: 3306, host: 8306, auto_correct: true
```

3) On your host machine run: `vagrant reload`.

4) As root on your Vagrant machine edit /etc/mysql/my.cnf and comment out the line that starts with "skip-external-locking" and the following like that starts with "bind-address".  More precisely:
```
43c43
> # skip-external-locking
---
< skip-external-locking
47c47
> # bind-address		= 127.0.0.1
---
< bind-address		= 127.0.0.1
```

5) Restart mysql on your Vagrant box by running:

`sudo /etc/init.d/mysql restart`

6) Start the webserver on the vagrant box with `runserver`.  You may want to make a change at this point to the database so you can test it later on.  For example, change the name of startup 10000 to "Foo, Inc.".

7) On your host machine, test that you can access the database by running:

`mysql -u masschallenge -psecret --port 8306 -h 0.0.0.0 mc_dev`

You should be able to see the change you made.  E.g.,

`select name from mc_startup where id=10000;`

8) On your host machine figure out your IP address either by following the instructions here:

http://osxdaily.com/2010/11/21/find-ip-address-mac/

or by running the shell command:

`ifconfig | grep netmask | tail -1 | sed 's/.*inet //' | sed 's/ .*//'`

9) In your impact-api repo edit .dev.env and change the DATABASE_URL to point to Vagrant database by going through the host machine's IP and using the masschallenge database account (password is 'secret').  More precisely:
```
5c5
< DATABASE_URL=mysql://root:root@mysql/mc_dev
---
> DATABASE_URL=mysql://masschallenge:secret@W.X.Y.Z:8306/mc_dev
```
where `W.X.Y.Z` is replaced by your host machine's IP address.

10) In your impact-api repo, make sure you are on the AC-4501 branch and run `make clean; make dev`.  This must be rerun to ensure that necessary migrations run on your local accelerate database.

11) In your impact-api repo, run:

`make grant-permissions PERMISSION_USER=37891-user@example.com PERMISSION_CLASSES=v1_clients`

(or the email of whatever user you want to use in Postman) and answer 'y' when it asks you if you want to create the group 'v1_clients'.

12) Login to the impact-api webserver as the user (37891-user@example.com) and create an application client.

13) In Postman, update the client id (and username/password if you haven't already).

14) In Postman, run 'POST oauth token'.  If you made a change, you should be able to see it either through Postman or with Browsable API.  For example, if you changed startup 10000, you should be able to see the change by going to 'startup list' in Postman.  If you do this, you will need to either be working with a superuser account or you will need to grant the permission `view_startup` (replace 'v1_clients' with 'view_startup' in step 11).

15) Run 'v1 user post'.  Assuming the user is successfully created, you will get back the new user's id.

16) Go to your local accelerate webpage and go through the password reset workflow for 'remerson@walden.pond'.  Specifically, goto login page, click "I forgot my password", enter 'remerson@walden.pond', check the webserver output for email with the reset link, go to that link after replacing the host name with your local server, and enter your new password.

17) Click 'LOGIN', enter 'remerson@walden.pond' and the password you just set.  You should be asked to accept our privacy policy and then you'll be taken to a member profile page. Rejoice!

**IMPORTANT**: After this testing you should disconnect the two environments by running the following:

In your accelerate repo:
% git checkout Vagrantfile
% vagrant reload

You may want to uncomment the lines in my.cnf and restart mysql, but it is not critical since it doesn't have a risk of getting checked in, the port is shutdown by the actions above, and it will all get cleaned up when you rebuild the vagrant environment.

In your impact-api repo:
% git checkout .dev.env
% make clean
